### PR TITLE
Add QgsDataProvider::uri property

### PR DESCRIPTION
**Pull request by m-kuhn**
Mainly for the purpose of not having to create this manually every single time it's used in a plugin for something.

But it might as well be saved as such to reduce the parsing overhead.

### CODE REVIEW

**Comment by 3nids** [comment]
editing sip files without auto-sipifying is not allowed anymore ;)

**Comment by m-kuhn** [comment]
Looks like this sipify bricolage doesn't handle nested curly brackets ;)

```
+    QString dataSourceUri( bool expandAuthConfig = false ) const;
+%Docstring
+ Get the data source specification. This may be a path or database
+ connection string
+ \param expandAuthConfig Whether to expand any assigned authentication configuration
+ :return: data source specification
+.. note::
+
+   The default authentication configuration expansion is FALSE. This keeps credentials
+ out of layer data source URIs and project files. Expansion should be specifically done
+ only when needed within a provider
+ :rtype: str
+%End
+      else
+      {
+        return mDataSourceURI;
+      }
+    }
```

**Comment by 3nids** [comment]
fixed

File: src/core/qgsdataprovider.h [hunk 1]
```diff
@@ -117,6 +134,25 @@ class CORE_EXPORT QgsDataProvider : public QObject
...
+    {
+      mDataSourceURI = uri.uri( true );
+    }
+
+    /**
+     * Get the data source specification.
+     *
+     * \since QGIS 3.0
+     */
+    QgsDataSourceUri uri()
```
Comment (nyalldawson) [inline src/core/qgsdataprovider.h]: const?

File: src/core/qgsdataprovider.h [hunk 2]
```diff
@@ -74,7 +91,10 @@ class CORE_EXPORT QgsDataProvider : public QObject
       CustomData   = 3000          //!< Custom properties for 3rd party providers or very provider-specific properties which are not expected to be of interest for other providers can be added starting from this value up.
     };
 
-    QgsDataProvider( QString const &uri = "" )
+    /**
+     * Create a new dataprovider with the specified in the \a uri.
```
Comment (agiudiceandrea) [inline src/core/qgsdataprovider.h:95]: Hi @m-kuhn, is something missing in this sentence?

Comment (m-kuhn) [reply src/core/qgsdataprovider.h:95]: Should we just drop the "in the"?

Comment (agiudiceandrea) [reply src/core/qgsdataprovider.h:95]: ...or "with the data source specification defined in the uri"?

Comment (m-kuhn) [reply src/core/qgsdataprovider.h:95]: Whatever you think adds the most information for a reader :+1:

**Pull request merged**